### ISSUE TITLE
Memory leak in ZStream#merge and ZStream#forever

### ISSUE THREAD

**Issue by fillson-shady** [zio/zio#3679]
Hello!

I faced huge memory leak in [scalapb/zio-grpc](https://github.com/scalapb/zio-grpc/) library (it uses ZIO RC-18-2) - millions of FiberContext's are spawned (and not garbage collected) because of using Stream.merge method. Here is the [code](https://github.com/scalapb/zio-grpc/blob/master/core/jvm/src/main/scala/scalapb/zio_grpc/client/ClientCalls.scala#L139).
I've extracted problematic code into a test and checked against newest ZIO versions (both RC-19-2 and SNAPSHOT). The problem with fibers leak has gone!

However now there is still another (very small) leak. Leaked objects are:

- zio.ZRef$Atomic
- zio.ZManaged$ReleaseMap$Running$1

 and some of ZIO tags.

Below is the code of the test (it doesn't completes nor outputs anything, so should be explored with the VisualVM):

```scala
import zio._
import zio.console._
import zio.duration._
import zio.stream._

object Test extends App {

  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] =
    for {

      queue <- Queue.bounded[Unit](requestedCapacity = 4096)

      streamIn = ZStream.fromQueue(queue)
      streamOut = ZStream.unit.forever.drain

      _ <- queue.offer(()).repeat(Schedule.forever).fork
      _ <- streamOut
        .merge(streamIn)
        .runDrain

    } yield 0
}
```

I wonder can it be connected with the #3488?

Thank you!

**Comment by iravid** [comment]
Yes, this is #3488. Thank you for reporting @fillson-shady. Keeping this open as this is a good reproducer.

**Comment by iravid** [comment]
These leaks should now be fixed by #3681 and #3691.
